api/controllers/emailSubscribe.js

```diff
@@ -59,177 +59,231 @@ exports.publicHead = async function (args, res) {
   return Actions.sendResponse(res, 200, data);
 };
 
+/**
+ * Fetch project names from the database, maintaining the order of provided IDs.
+ * 
+ * @param {Model} ProjectModel Mongoose Project model
+ * @param {ObjectId[]} projectIds Array of project ObjectIds
+ * @param {string} fallbackName Optional fallback name if no projects found
+ * @returns {Promise<string[]>} Array of project names in the same order as input IDs
+ */
+const fetchProjectNames = async (ProjectModel, projectIds, fallbackName) => {
+  if (!Array.isArray(projectIds) || projectIds.length === 0) {
+    return fallbackName ? [fallbackName] : [];
```

Comment:
It could be quicker to add a default value to the fallbackName argument and skip these checks, assuming fallbackName is a generic value. If it's dynamic then this works.